### PR TITLE
Validate all example and sample applications

### PR DESCRIPTION
**Pull request by marcospereira**
### What changes were proposed in this pull request?

Some of them were not compiled or tested. The script now discovers all apps and tests them.

### Why are the changes needed?

To make sure all apps are in good shape, compiling, and building as expected.

### Does this PR introduce any user-facing change?

No API changes at all.

### How was this patch tested?

Running `./scripts/build-all.sh test` so far.

### CODE REVIEW

File: examples/snippets/modules/ROOT/examples/akkastreams-grpc-java/src/main/java/sensordata/Logger.java [hunk 1]
```diff
@@ -5,16 +5,18 @@
...
 import sensordata.grpc.SensorData;
 
 public class Logger extends AkkaStreamlet {
-    private final ProtoInlet<SensorData> inlet = new ProtoInlet<SensorData>(
+    private final ProtoInlet<SensorData> inlet = new ProtoInlet<>(
             "in",
             SensorData.class,
-            true
+            true,
+            CodecInlet::logAndSkip
```
Comment (marcospereira) [inline examples/snippets/modules/ROOT/examples/akkastreams-grpc-java/src/main/java/sensordata/Logger.java:19]: The code here was not compiling. I'm not sure this is the correct fix for it, though.

File: examples/snippets/modules/ROOT/examples/base-images-override/build.sbt [hunk 2]
```diff
@@ -3,8 +3,9 @@ import sbt.Keys._
 
     //tag::docs-projectSetup-example[]
 lazy val sampleApp = (project in file("."))
+    .enablePlugins(CloudflowApplicationPlugin)
     .settings(
-      cloudflowAkkaBaseImage := "myRepositoryUrl/myRepositoryPath:2.0.10-cloudflow-akka-2.6.6-scala-2.12",
+      cloudflowDockerBaseImage := "myRepositoryUrl/myRepositoryPath:2.0.10-cloudflow-akka-2.6.6-scala-2.12",
```
Comment (marcospereira) [inline examples/snippets/modules/ROOT/examples/base-images-override/build.sbt:8]: I'm not sure if there are other places in docs where we should recommend the new sbt setting instead of the deprecated ones.

File: scripts/build-all.sh [hunk 3]
```diff
@@ -25,16 +25,16 @@ then
...
   echo "==================================================================================="
-  exit -1
+  exit 1
 fi
 echo "========================================================================="
 echo "Runs 'sbt $TARGET' for core and examples"
 echo "========================================================================="
 
 cd $DIR/../core
-sbt -mem 2048 --supershell=false "; scalafmtCheck ; $TARGET  ; +publishLocal"
```
Comment (marcospereira) [inline scripts/build-all.sh:35]: I got an OOM error when running locally. Maybe this is a too aggressive bump?

File: scripts/build-all.sh [hunk 4]
```diff
@@ -45,24 +45,32 @@ echo "Now starting building of examples..."
...
 
 for prj in $PROJECTS; do
   echo "========================================================================="
   echo "${TARGET}: $prj"
   echo "========================================================================="
 
-  cd $prj
+  pushd "$prj"
+
+  # Only run `verifyBlueprint` task if there are blueprint files in the project
```
Comment (marcospereira) [inline scripts/build-all.sh:58]: Not all examples have a `blueprint.conf` file.

**Review by marcospereira — COMMENTED** [review]
This was sitting on my machine since yesterday's night. There are some other minor changes to make all the samples run in the CI. I will push them when I have the time.

File: examples/snippets/modules/ROOT/examples/akkastreams-grpc-java/src/main/java/sensordata/Logger.java (hunk 1 above)
Comment (RayRoestenburg) [reply examples/snippets/modules/ROOT/examples/akkastreams-grpc-java/src/main/java/sensordata/Logger.java:19]: There is a Java API for this, `ProtoInlet.create` (it automatically uses `logAndSkip`), which is probably better for the example.

Comment (marcospereira) [reply examples/snippets/modules/ROOT/examples/akkastreams-grpc-java/src/main/java/sensordata/Logger.java]: Resolved when rebasing with the main branch.

File: .github/workflows/build-pr.yaml [hunk 5]
```diff
@@ -2,34 +2,30 @@ name: Build PRs
...
+    
     - name: Setup java
       uses: joschi/setup-jdk@v2
       with:
         java-version: ${{ matrix.java }}
         architecture: x64
+    
     - name: Cache SBT and Coursier cache
       uses: coursier/cache-action@v3
-    - name: Scalafmt check
```
Comment (marcospereira) [inline .github/workflows/build-pr.yaml:20]: This was duplicated with `push.yaml`, which also runs for pull requests.

File: .github/workflows/build-pr.yaml [hunk 6]
```diff
@@ -2,34 +2,30 @@ name: Build PRs
...
-      run: |
-        cd core
-        sbt "scalafmtCheckAll; scalafmtSbtCheck"
-    - name: Build documentation
```
Comment (marcospereira) [inline .github/workflows/build-pr.yaml:24]: This was duplicated with `push.yaml`, which also runs for pull requests.

File: .github/workflows/build-pr.yaml [hunk 7]
```diff
@@ -2,34 +2,30 @@ name: Build PRs
...
-        cd docs
-        make html-author-mode
-    - name: Run integration tests
+    
+    - name: Run scripted tests
       run: |
         git config --global user.email "cloudflow@lightbend.com"
         git config --global user.name "Cloudflow CI"
         cd core
-        sbt -mem 2048 +test
```
Comment (marcospereira) [inline .github/workflows/build-pr.yaml:34]: `core/tests` runs for `push.yaml` too.

File: .github/workflows/push.yml [hunk 8]
```diff
@@ -23,23 +23,65 @@ jobs:
...
       env:
         SBT_OPTS: -Xms512M -Xmx2048M -Xss2M -XX:MaxMetaspaceSize=1024M
-      run: |
-        export PATH="$HOME/opt/sbt/bin:$PATH"
-        export SBT_OPTS="-Xss256M"
-        cd scripts
-        ./build-all.sh +test
+      run: ./scripts/build-core.sh +test
+
+  build-and-test-sbt-examples:
```
Comment (marcospereira) [inline .github/workflows/push.yml:51]: Split into different jobs so that they can run independently.

File: .github/workflows/push.yml [hunk 9]
```diff
@@ -23,23 +23,65 @@ jobs:
...
+    name: "sbt samples"
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        java: [ '8', '11', '13' ]
```
Comment (marcospereira) [inline .github/workflows/push.yml:57]: Perhaps we don't want these samples to run on all these JDK versions.

File: core/build.sbt [hunk 10]
```diff
@@ -176,6 +176,11 @@ setVersionFromTag := {
     .getOrElse("0.0.0-SNAPSHOT")}"""")
 }
 
+lazy val writeVersionToFile = taskKey[Unit]("Write project version to file")
```
Comment (marcospereira) [inline core/build.sbt:179]: This helps to have a consistent version that will be used when running the tests for all the samples. The file will later be read by the build shell scripts.

The goal here is to avoid tasks that change files to break dynver (used as a fallback by the samples) since it generates another version if the git repository is dirty.

**Comment by RayRoestenburg** [comment]
Sure, let's look at these as possible improvements and pick it up in a support cycle. @andreaTP might also be interested in these changes

File: .github/workflows/build-pr.yaml [hunk 11]
```diff
@@ -2,34 +2,30 @@ name: Build PRs
 on:
   pull_request:
 jobs:
-  build-all:
+  run-scripted-tests:
     runs-on: ubuntu-latest
     strategy:
+      fail-fast: false
```
Comment (andreaTP) [inline .github/workflows/build-pr.yaml:8]: I don't think we want to avoid the `fail-fast` semantic here

File: .github/workflows/build-pr.yaml (hunk 7 above)
Comment (andreaTP) [reply .github/workflows/build-pr.yaml:34]: I agree on removing the duplicated `scalafmt` and `docs` build, but the `+test` is intentional and guarantees compatibility on the different JVMs

File: .github/workflows/build-test-tools.yml [hunk 12]
```diff
@@ -6,14 +6,13 @@ on:
     branches: [main]
 
 jobs:
-  build:
-    runs-on: ubuntu-18.04
+  build-binary:
+    runs-on: ubuntu-latest
```
Comment (andreaTP) [inline .github/workflows/build-test-tools.yml:10]: is this change needed, we should prefer stable identifiers

File: .github/workflows/build_and_publish.yaml [hunk 13]
```diff
@@ -8,8 +8,7 @@ jobs:
   build-docs:
     runs-on: ubuntu-latest
     steps:
-    - name: Checkout
-      uses: actions/checkout@v2.3.3
+    - uses: actions/checkout@v2
```
Comment (andreaTP) [inline .github/workflows/build_and_publish.yaml:11]: Have you tested this workflow in isolation to verify this change?
I'm pretty sure we experienced a few regressions on the versioning here ...

File: .github/workflows/push.yml [hunk 14]
```diff
@@ -23,23 +23,65 @@ jobs:
...
     steps:
-    - uses: actions/checkout@v2
 
-    - name: Setup environment
-      run: |
-        make -C .github/workflows/setup setup-local-all
```
Comment (andreaTP) [inline .github/workflows/push.yml:33]: good catch we can remove the entire `Makefile`:
https://github.com/lightbend/cloudflow/blob/main/.github/workflows/setup/Makefile

File: .github/workflows/push.yml [hunk 15]
```diff
@@ -23,23 +23,65 @@ jobs:
...
-      run: |
-        export PATH="$HOME/opt/sbt/bin:$PATH"
-        export SBT_OPTS="-Xss256M"
-        cd scripts
-        ./build-all.sh +test
+      run: ./scripts/build-core.sh +test
+
+  build-and-test-sbt-examples:
+    name: "sbt samples"
+    runs-on: ubuntu-latest
```
Comment (andreaTP) [inline .github/workflows/push.yml:53]: we should prefer fixed versions

File: .github/workflows/push.yml (hunk 14 above)
Comment (andreaTP) [reply .github/workflows/push.yml:33]: oh, it's done, nice, thanks!

File: core/build.sbt (hunk 10 above)
Comment (andreaTP) [reply core/build.sbt:179]: Please remove this customization.

> The goal here is to avoid tasks that change files to break dynver

That's the reason why we should keep it as it was, the tasks should not change files to while running, otherwise, we want the CI to fail.

File: core/cloudflow-it/swiss-knife/project/cloudflow-plugins.sbt [hunk 16]
```diff
@@ -1,5 +1,5 @@
 val latestVersion = {
-  sys.env.get("CLOUDFLOW_VERSION").fold(
+  sys.env.get("CLOUDFLOW_VERSION").filter(_.nonEmpty).fold(
```
Comment (andreaTP) [inline core/cloudflow-it/swiss-knife/project/cloudflow-plugins.sbt:2]: have you done a global search and replace for this?

File: examples/snippets/modules/ROOT/examples/base-images-override/.scalafmt.conf [hunk 17]
```diff
@@ -0,0 +1,44 @@
+version = "2.3.2"
```
Comment (andreaTP) [inline examples/snippets/modules/ROOT/examples/base-images-override/.scalafmt.conf:1]: is the `scalafmt` configuration all the same across the repo?
was, this one, missing?

File: examples/snippets/modules/ROOT/examples/base-images-override/build.sbt (hunk 2 above)
Comment (andreaTP) [inline examples/snippets/modules/ROOT/examples/base-images-override/build.sbt:8]: I'm not sure what is going on here ...

File: examples/snippets/modules/ROOT/examples/base-images-override/project/build.properties [hunk 18]
```diff
@@ -0,0 +1 @@
+sbt.version = 1.4.4
```
Comment (andreaTP) [inline examples/snippets/modules/ROOT/examples/base-images-override/project/build.properties:1]: the rest of the repo standardize on `1.4.9`

File: scripts/functions.sh [hunk 19]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
+
+SCRIPTS_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd -P )"
+ROOT_DIR=$(dirname "$SCRIPTS_DIR")
+
+# Generates a stable version that will be used instead of reading 
+# from dynver. This is useful when running tasks that can change the
+# code for all the samples, for example, when running format tools.
+function echo_cf_version() {
```
Comment (andreaTP) [inline scripts/functions.sh:9]: if there are issues with dynver, let's analyze them and agree on solutions, but this should not be done in this way.

File: scripts/build-mvn-examples.sh [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+
+for prj in "${PARENT_PROJECTS[@]}"; do
+  show_message "mvn ${TARGET}: $prj"
+  cd "$prj"
+
+  # Only run `cloudflow:verify-blueprint` task if there are blueprint files in the project
+  if [[ -n $(find . -name "blueprint.conf") ]]; then
+    # For more details about Maven workflow, see the docs:
+    # https://developer.lightbend.com/docs/cloudflow/shared-content/2.1.2/develop/maven-support.html
+    if ! mvn package cloudflow:extract-streamlets -DskipTests; then
```
Comment (andreaTP) [inline scripts/build-mvn-examples.sh:57]: Can we remove `-DskipTests` ?

**Review by andreaTP — CHANGES_REQUESTED** [review]
I will review this better tomorrow, overall I really like the separation of the bash scripts, thanks @marcospereira !

The current issue with this PR is how it plays with `dynver` we should be able to avoid fixing the version.

File: .github/workflows/build-pr.yaml (hunk 7 above)
Comment (marcospereira) [reply .github/workflows/build-pr.yaml:34]: Should we make the change there instead? I mean, make this run for multiple JVMs:

https://github.com/lightbend/cloudflow/blob/e9a48e29630df7f9629108539ed996d14eb89184/.github/workflows/push.yml#L26-L45

File: core/build.sbt (hunk 10 above)
Comment (marcospereira) [reply core/build.sbt:179]: It will fail anyway, but instead of failing with unresolved artifacts for the next sample, it would fail for the right reason if that is the case.

This is also relevant when running the Maven examples where we need the `CLOUDFLOW_VERSION` env since we don't have dynver there. Or is there an option for Maven samples?

File: .github/workflows/build-pr.yaml (hunk 11 above)
Comment (marcospereira) [reply .github/workflows/build-pr.yaml]: Fixed here: 637ca491f5b3efe7457d9186e6187d8ae9179e4d

File: .github/workflows/build_and_publish.yaml (hunk 13 above)
Comment (marcospereira) [reply .github/workflows/build_and_publish.yaml:11]: I have not. How can I do that?

File: core/cloudflow-it/swiss-knife/project/cloudflow-plugins.sbt (hunk 16 above)
Comment (marcospereira) [reply core/cloudflow-it/swiss-knife/project/cloudflow-plugins.sbt:2]: Yes. :-)

File: examples/snippets/modules/ROOT/examples/base-images-override/.scalafmt.conf (hunk 17 above)
Comment (marcospereira) [reply examples/snippets/modules/ROOT/examples/base-images-override/.scalafmt.conf:1]: They are not the same across the repo, but I'm pretty sure they are the same for all the samples. This one was copied from one of the other samples.

File: examples/snippets/modules/ROOT/examples/base-images-override/build.sbt (hunk 2 above)
Comment (marcospereira) [reply examples/snippets/modules/ROOT/examples/base-images-override/build.sbt:8]: Avoid using the deprecated setting in the example.

File: examples/snippets/modules/ROOT/examples/base-images-override/project/build.properties (hunk 18 above)
Comment (marcospereira) [reply examples/snippets/modules/ROOT/examples/base-images-override/project/build.properties:1]: The vast majority of the examples are using 1.4.4, and this was probably just following one of them. I thought about changing all of them to use `1.4.9`, but this PR already has too many changes (this is why I've started https://github.com/lightbend/cloudflow/pull/1075).

File: scripts/functions.sh (hunk 19 above)
Comment (marcospereira) [reply scripts/functions.sh:9]: There is no issue with dynver: if the repo is dirty it will generate a version that contains the time.

File: scripts/build-mvn-examples.sh (hunk 20 above)
Comment (marcospereira) [reply scripts/build-mvn-examples.sh:57]: Tests will run later (see a little below where we run the target), so we don't need to run them here.

File: .github/workflows/build-pr.yaml (hunk 7 above)
Comment (andreaTP) [reply .github/workflows/build-pr.yaml:34]: I have no strong opinions here, I think I'll prefer it here, but I won't disagree on moving it to `push.yml`

File: .github/workflows/push.yml (hunk 9 above)
Comment (andreaTP) [reply .github/workflows/push.yml:55]: agreed samples can run on java 8 only

File: core/build.sbt (hunk 10 above)
Comment (andreaTP) [reply core/build.sbt:179]: for Maven something like:
```bash
export CLOUDFLOW_VERSION=$(sbt -no-colors --error 'print version')
```
should make it work.

> It will fail anyway

If the git repo is not dirty the version is always the same.
But maybe I'm misunderstanding, can you elaborate? or maybe is faster to go through in a sync meeting?

File: .github/workflows/build_and_publish.yaml (hunk 13 above)
Comment (andreaTP) [reply .github/workflows/build_and_publish.yaml:11]: you can hack the workflow in a branch to be triggered on `push` and verify that the output (e.g. the version) is correct and cosistent

File: examples/snippets/modules/ROOT/examples/base-images-override/.scalafmt.conf (hunk 17 above)
Comment (andreaTP) [reply examples/snippets/modules/ROOT/examples/base-images-override/.scalafmt.conf:1]: thanks!

File: examples/snippets/modules/ROOT/examples/base-images-override/project/build.properties (hunk 18 above)
Comment (andreaTP) [reply examples/snippets/modules/ROOT/examples/base-images-override/project/build.properties:1]: fine with it

File: scripts/functions.sh (hunk 19 above)
Comment (andreaTP) [reply scripts/functions.sh:9]: correct, and building/running examples should not result in a dirty repo

File: scripts/build-sbt-examples.sh [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+
+  case "$prj" in
+    *-java)
+      if ! sbt -mem 4096 --supershell=false "; $TARGET; $blueprintTask"; then
+        show_message "Failed to run $TARGET and other checks for $prj"
+        exit 1
+      fi
+      ;;
+    *)
+      if ! sbt -mem 4096 --supershell=false "; scalafmtCheckAll ; $TARGET; $blueprintTask"; then
```
Comment (andreaTP) [inline scripts/build-sbt-examples.sh:54]: minor, can we append `scalafmtCheckAll` the same way as `blueprintTask`? I find it much cleaner.

Comment (marcospereira) [reply scripts/build-sbt-examples.sh:54]: Do you mean to run `scalafmtCheckAll` only if there is a `.scalafmt.conf` file?

Comment (andreaTP) [reply scripts/build-sbt-examples.sh:54]: I mean using this pattern:
```bash
  # Only run `verifyBlueprint` task if there are blueprint files in the project
  blueprintTask=""
  if [[ -n $(find . -name "blueprint.conf") ]]; then
    blueprintTask="verifyBlueprint"
  fi
```

and removing the `  case "$prj" in`

Comment (marcospereira) [reply scripts/build-sbt-examples.sh]: Done. See https://github.com/lightbend/cloudflow/pull/1059/commits/0a60944d528585bbbb4df9367037ca4319391c00.

File: scripts/functions.sh (hunk 19 above)
Comment (marcospereira) [reply scripts/functions.sh:6]: Done. See https://github.com/lightbend/cloudflow/pull/1059/commits/2b356c499ba8328da59804f1a05f3f174b133337.

Comment (andreaTP) [reply scripts/functions.sh:6]: nice 👍

File: scripts/build-sbt-examples.sh (hunk 21 above)
Comment (andreaTP) [reply scripts/build-sbt-examples.sh]: nice 👍

**Review by andreaTP — APPROVED** [review]
Added one commit to test examples only on Java 8.

Thanks for the hard work @marcospereira ! LGTM!
